File: apps/api/app/models/redaction.py
"""SQLAlchemy model for AI document redaction jobs."""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.models.base import ModelMixin


class RedactionJob(Base, ModelMixin):
    """Tracks a single PII-detection + redaction workflow for a document.

    Lifecycle: pending → analyzing → review → applying → done | failed
    """

    __tablename__ = "redaction_jobs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        server_default="gen_random_uuid()",
    )
    org_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    document_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )
    # pending | analyzing | review | applying | done | failed
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="pending", server_default="pending"
    )
    # List of detected PII entity dicts (each includes id, entity_type, text, page, confidence, position, is_high_sensitivity)
    detected_entities: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    # Subset of detected_entities approved by the user for actual redaction
    approved_redactions: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    # UUID of the newly created redacted document in the data room (nullable until done)
    redacted_document_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )
    redacted_s3_key: Mapped[str | None] = mapped_column(String(1024), nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    entity_count: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default="0"
    )
    approved_count: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default="0"
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, server_default="now()"
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, server_default="now()"
    )

File: apps/web/src/lib/redaction.ts
/**
 * AI Document Redaction — types and React Query hooks.
 *
 * Mirrors the FastAPI Pydantic schemas defined in
 * apps/api/app/modules/redaction/schemas.py
 */

import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { api } from "@/lib/api";

// ── Types ───────────────────────────────────────────────────────────────────

export type RedactionStatus =
  | "pending"
  | "analyzing"
  | "review"
  | "applying"
  | "done"
  | "failed";

export interface DetectedEntity {
  id: number;
  entity_type: string;
  text: string;
  page: number;
  confidence: number;
  position: { x: number; y: number; width: number; height: number };
  is_high_sensitivity: boolean;
}

export interface RedactionJob {
  id: string;
  document_id: string;
  status: RedactionStatus;
  detected_entities: DetectedEntity[] | null;
  approved_redactions: DetectedEntity[] | null;
  entity_count: number;
  approved_count: number;
  redacted_document_id: string | null;
  redacted_s3_key: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface EntityTypeInfo {
  entity_type: string;
  is_high_sensitivity: boolean;
}

export interface RedactionRules {
  entity_types: EntityTypeInfo[];
  high_sensitivity_types: string[];
}

export interface AnalyzeDocumentRequest {
  document_text?: string;
}

export interface ApproveRedactionsRequest {
  approved_entity_ids: number[];
}

// ── Query keys ───────────────────────────────────────────────────────────────

const KEYS = {
  jobs: (documentId?: string) =>
    documentId
      ? (["redaction", "jobs", documentId] as const)
      : (["redaction", "jobs"] as const),
  job: (jobId: string) => ["redaction", "job", jobId] as const,
  rules: () => ["redaction", "rules"] as const,
};

// ── Hooks ────────────────────────────────────────────────────────────────────

/** List redaction jobs, optionally scoped to a single document. */
export function useRedactionJobs(documentId?: string) {
  return useQuery<RedactionJob[]>({
    queryKey: KEYS.jobs(documentId),
    queryFn: async () => {
      const params = documentId ? { document_id: documentId } : {};
      const { data } = await api.get("/redaction/jobs", { params });
      return data;
    },
  });
}

/** Fetch a single redaction job by ID, polling while it is in-progress. */
export function useRedactionJob(jobId: string, enabled = true) {
  return useQuery<RedactionJob>({
    queryKey: KEYS.job(jobId),
    queryFn: async () => {
      const { data } = await api.get(`/redaction/jobs/${jobId}`);
      return data;
    },
    enabled: enabled && !!jobId,
    refetchInterval: (query) => {
      const status = query.state.data?.status;
      // Poll while the job is still running
      if (status === "pending" || status === "analyzing" || status === "applying") {
        return 2000; // 2 s
      }
      return false;
    },
  });
}

/** Return the list of entity types and which are high-sensitivity. */
export function useRedactionRules() {
  return useQuery<RedactionRules>({
    queryKey: KEYS.rules(),
    queryFn: async () => {
      const { data } = await api.get("/redaction/rules");
      return data;
    },
    staleTime: Infinity, // static data
  });
}

/** Create a redaction job and queue PII analysis. Returns { job_id, status }. */
export function useAnalyzeDocument() {
  const qc = useQueryClient();
  return useMutation<
    { job_id: string; status: string },
    Error,
    { document_id: string; document_text?: string }
  >({
    mutationFn: async ({ document_id, document_text }) => {
      const { data } = await api.post(
        `/redaction/analyze/${document_id}`,
        document_text ? { document_text } : {}
      );
      return data;
    },
    onSuccess: (_data, variables) => {
      qc.invalidateQueries({ queryKey: KEYS.jobs(variables.document_id) });
      qc.invalidateQueries({ queryKey: KEYS.jobs() });
    },
  });
}

/** Approve a subset of detected entities for actual redaction. */
export function useApproveRedactions() {
  const qc = useQueryClient();
  return useMutation<
    RedactionJob,
    Error,
    { job_id: string; approved_entity_ids: number[] }
  >({
    mutationFn: async ({ job_id, approved_entity_ids }) => {
      const { data } = await api.post(`/redaction/jobs/${job_id}/approve`, {
        approved_entity_ids,
      });
      return data;
    },
    onSuccess: (data) => {
      qc.setQueryData(KEYS.job(data.id), data);
      qc.invalidateQueries({ queryKey: KEYS.jobs(data.document_id) });
    },
  });
}

/** Trigger PDF generation with redaction boxes applied. */
export function useApplyRedaction() {
  const qc = useQueryClient();
  return useMutation<RedactionJob, Error, { job_id: string }>({
    mutationFn: async ({ job_id }) => {
      const { data } = await api.post(`/redaction/jobs/${job_id}/apply`);
      return data;
    },
    onSuccess: (data) => {
      qc.setQueryData(KEYS.job(data.id), data);
      qc.invalidateQueries({ queryKey: KEYS.jobs(data.document_id) });
    },
  });
}
